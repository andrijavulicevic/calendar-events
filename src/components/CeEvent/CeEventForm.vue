<template>
  <v-card class="elevation-12">
    <v-toolbar color="primary" dark flat>
      <v-toolbar-title>{{ formName }}</v-toolbar-title>
      <v-spacer></v-spacer>
      <v-btn icon @click="$emit('close')">
        <v-icon>mdi-close</v-icon>
      </v-btn>
    </v-toolbar>
    <v-card-text>
      <v-form>
        <v-text-field
          v-model="calendarEvent.name"
          v-validate="'required'"
          :error-messages="errors.collect('calendarEvent.name')"
          label="Name"
          data-vv-name="calendarEvent.name"
          data-vv-as="name"
          required
        ></v-text-field>

        <v-layout justify-space-between>
          <v-flex md5>
            <CeDatePicker
              label="Start date"
              :initialDate="chosenStart.date"
              @dateChanged="setStartDate"
            />
          </v-flex>
          <v-flex md5>
            <CeDatePicker
              label="End date"
              :initialDate="chosenEnd.date"
              :minDate="chosenEnd.minDate"
              @dateChanged="setEndDate"
            />
          </v-flex>
        </v-layout>

        <v-layout justify-space-between v-if="!allDayEvent">
          <v-flex md5>
            <CeTimePicker
              label="Start time"
              :initialTime="chosenStart.time"
              @timeChanged="setStartTime"
            />
          </v-flex>
          <v-flex md5>
            <CeTimePicker
              label="End time"
              :initialTime="chosenEnd.time"
              @timeChanged="setEndTime"
            />
          </v-flex>
        </v-layout>

        <v-checkbox v-model="allDayEvent" label="All day event"></v-checkbox>

        <v-select
          v-model="calendarEvent.visibility"
          :items="visibilityOptions"
          label="Visibility"
        ></v-select>

        <v-textarea
          v-model="calendarEvent.details"
          v-validate="'required'"
          :error-messages="errors.collect('calendarEvent.details')"
          label="Details"
          data-vv-name="calendarEvent.details"
          data-vv-as="details"
          rows="3"
        ></v-textarea>

        <v-select
          v-show="isPrivateEvent"
          v-model="calendarEvent.participants"
          :items="participants"
          label="Participants"
          multiple
        ></v-select>

        <v-text-field
          v-model="calendarEvent.location"
          v-validate="'required'"
          :error-messages="errors.collect('calendarEvent.location')"
          label="Location"
          data-vv-name="calendarEvent.location"
          data-vv-as="location"
          required
        ></v-text-field>
      </v-form>
    </v-card-text>
    <v-card-actions>
      <v-spacer></v-spacer>
      <v-btn
        color="blue darken-1"
        text
        :loading="loading"
        @click="executeFormAction"
        >{{ formActionName }}</v-btn
      >
    </v-card-actions>
  </v-card>
</template>

<script>
import { mapGetters } from "vuex";
import { addDays, isAfter } from "date-fns";
import {
  GET_EVENTS_LOADING,
  GET_ALL_PARTICIPANTS
} from "../../store/getters.type";
import { concatDateAndTime } from "../../utils/date-utils";
import CeDatePicker from "./CeDatePicker";
import CeTimePicker from "./CeTimePicker";

export default {
  props: {
    formName: {
      type: String,
      required: true
    },
    formActionName: {
      type: String,
      required: true
    },
    initialFormData: {
      type: Object
    }
  },
  components: {
    CeDatePicker,
    CeTimePicker
  },
  created() {
    if (this.initialFormData) {
      this.calendarEvent = {
        ...this.calendarEvent,
        ...this.initialFormData
      };
      if (this.initialFormData.participants) {
        this.calendarEvent.participants = [
          ...this.initialFormData.participants
        ];
      }
    }
  },
  mounted() {
    const [startDate, startTime] = this.calendarEvent.start.split(" ");
    const [endDate, endTime] = this.calendarEvent.end.split(" ");
    this.chosenStart.date = startDate;
    this.chosenEnd.date = endDate;
    this.chosenEnd.minDate = this.calculateEndMinDate();
    if (startTime) {
      this.allDayEvent = false;
      this.chosenStart.time = startTime;
      this.chosenEnd.time = endTime;
    }
  },
  data: () => ({
    chosenStart: {
      date: "",
      time: null
    },
    chosenEnd: {
      date: "",
      minDate: "",
      time: null
    },
    allDayEvent: true,
    visibilityOptions: ["Private", "Public"],
    calendarEvent: {
      name: "",
      start: "",
      end: "",
      visibility: "Private",
      details: "",
      location: "",
      participants: [],
      color: "blue"
    }
  }),
  computed: {
    ...mapGetters({
      loading: GET_EVENTS_LOADING,
      participants: GET_ALL_PARTICIPANTS
    }),
    isPrivateEvent() {
      return this.calendarEvent.visibility === "Private";
    }
  },
  watch: {
    "chosenStart.date": {
      handler: function(newVal) {
        if (isAfter(new Date(newVal), new Date(this.chosenEnd.date))) {
          this.chosenEnd.date = newVal;
          this.chosenEnd.minDate = this.calculateEndMinDate();
        }
      }
    },
    allDayEvent(newVal) {
      if (!newVal) {
        if (!this.chosenStart.time) {
          this.chosenStart.time = "12:00";
          this.chosenEnd.time = "12:30";
        }
      } else {
        this.chosenStart.time = null;
        this.chosenEnd.time = null;
      }
    },
    isPrivateEvent(newVal) {
      if (!newVal) {
        this.calendarEvent.participants = [];
      }
    }
  },
  methods: {
    setStartDate(date) {
      this.chosenStart.date = date;
    },
    setEndDate(date) {
      this.chosenEnd.date = date;
    },
    setStartTime(time) {
      this.chosenStart.time = time;
    },
    setEndTime(time) {
      this.chosenEnd.time = time;
    },
    calculateEndMinDate() {
      return addDays(new Date(this.chosenStart.date), -1).toISOString();
    },
    executeFormAction() {
      this.$validator.validate().then(valid => {
        if (!valid) return;
        this.calendarEvent.start = concatDateAndTime(
          this.chosenStart.date,
          this.chosenStart.time
        );
        this.calendarEvent.end = concatDateAndTime(
          this.chosenEnd.date,
          this.chosenEnd.time
        );
        this.$emit("eventUpdated", this.calendarEvent);
      });
    }
  }
};
</script>
