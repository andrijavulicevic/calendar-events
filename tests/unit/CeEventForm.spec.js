import { createLocalVue, mount } from "@vue/test-utils";
import Vuetify from "vuetify";
import Vuex from "vuex";
import VeeValidate from "vee-validate";
import CeEventForm from "../../src/components/CeEvent/CeEventForm";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuetify);
localVue.use(VeeValidate);

let vuetify, wrapper, $store;
const propsData = {
  formName: "Create event",
  formActionName: "Create",
  initialFormData: {
    start: "2019-08-22",
    end: "2019-08-23"
  }
};

describe("CeEventForm.vue", () => {
  beforeEach(() => {
    vuetify = new Vuetify();
    $store = {
      getters: {
        getEventsLoading: false,
        getAllParticipants: []
      }
    };
    wrapper = mount(CeEventForm, {
      localVue,
      vuetify,
      propsData,
      mocks: { $store }
    });
  });

  it("is a Vue instance", () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("is an CeEventForm component", () => {
    expect(wrapper.is(CeEventForm)).toBeTruthy();
  });

  it("should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("CeEventForm props", () => {
    it("has 'formName' property", () => {
      expect(wrapper.props("formName")).toBeTruthy();
    });

    it("'formName' is of type String", () => {
      expect(typeof wrapper.props("formName")).toBe("string");
    });

    it("value of 'formName' is 'Create event'", () => {
      expect(wrapper.props("formName")).toBe("Create event");
    });

    it("has 'formActionName' property", () => {
      expect(wrapper.props("formActionName")).toBeTruthy();
    });

    it("'formActionName' is of type String", () => {
      expect(typeof wrapper.props("formActionName")).toBe("string");
    });

    it("vale of 'formActionName' is 'Create'", () => {
      expect(wrapper.props("formActionName")).toBe("Create");
    });

    it("has 'initialFormData' property", () => {
      expect(wrapper.props("initialFormData")).toBeTruthy();
    });

    it("'initialFormData' is of type Object", () => {
      expect(typeof wrapper.props("initialFormData")).toBe("object");
    });

    it("'initialFormData' prop has fields 'start' and 'end'", () => {
      expect(wrapper.props("initialFormData").start).toBeTruthy();
      expect(wrapper.props("initialFormData").end).toBeTruthy();
      expect(wrapper.props("initialFormData").start).toBe("2019-08-22");
      expect(wrapper.props("initialFormData").end).toBe("2019-08-23");
    });
  });

  describe("CeEventForm data", () => {
    it("has data", () => {
      expect(CeEventForm.data).toBeTruthy();
    });

    it("data is of type function", () => {
      expect(typeof CeEventForm.data).toBe("function");
    });

    it("'chosenStart' is defined and of type Object", () => {
      expect(wrapper.vm.chosenStart).toBeDefined();
      expect(typeof wrapper.vm.chosenStart).toBe("object");
    });

    it("'chosenStart' has initial value '{date: \"2019-08-22\", time: null}'", () => {
      expect(wrapper.vm.chosenStart.date).toBe("2019-08-22");
      expect(wrapper.vm.chosenStart.time).toBe(null);
    });

    it("'chosenEnd' is defined and of type Object", () => {
      expect(wrapper.vm.chosenEnd).toBeDefined();
      expect(typeof wrapper.vm.chosenEnd).toBe("object");
    });

    it("'chosenEnd' has initial value '{date: \"2019-08-23\", time: null}'", () => {
      expect(wrapper.vm.chosenEnd.date).toBe("2019-08-23");
      expect(wrapper.vm.chosenEnd.time).toBe(null);
    });

    it("'allDayEvent' is defined and of type Boolean", () => {
      expect(wrapper.vm.allDayEvent).toBeDefined();
      expect(typeof wrapper.vm.allDayEvent).toBe("boolean");
    });

    it("'allDayEvent' has initial value true", () => {
      expect(wrapper.vm.allDayEvent).toBe(true);
    });

    it("'visibilityOptions' is defined and of type Array", () => {
      expect(wrapper.vm.visibilityOptions).toBeDefined();
      expect(wrapper.vm.visibilityOptions instanceof Array).toBe(true);
    });

    it("'visibilityOptions' has length of 2", () => {
      expect(wrapper.vm.visibilityOptions.length).toBe(2);
    });

    it("'calendarEvent' is defined and of type Object", () => {
      expect(wrapper.vm.calendarEvent).toBeDefined();
      expect(typeof wrapper.vm.calendarEvent).toBe("object");
    });
  });

  describe("CeEventForm computed", () => {
    it("computed 'loading' is defined", () => {
      expect(wrapper.vm.loading).toBeDefined();
    });

    it("initial value of 'loading' is false", () => {
      expect(typeof CeEventForm.loading).toBe("boolean");
      expect(wrapper.vm.isPrivateEvent).toBe(false);
    });

    it("computed 'participants' is defined", () => {
      expect(wrapper.vm.participants).toBeDefined();
    });

    it("initial value of 'participants' is empty array", () => {
      expect(CeEventForm.participants instanceof Array).toBe(true);
      expect(wrapper.vm.participants.length).toBe(0);
    });

    it("computed 'isPrivateEvent' is defined", () => {
      expect(wrapper.vm.isPrivateEvent).toBeDefined();
    });

    it("initial value of 'isPrivateEvent' is true", () => {
      expect(wrapper.vm.isPrivateEvent).toBe(true);
    });

    it("value of 'isPrivateEvent' changes if 'calendarEvent.visibility changes'", () => {
      wrapper.setData({ calendarEvent: { visibility: "Public" } });
      expect(wrapper.vm.isPrivateEvent).toBe(false);
    });
  });
});
