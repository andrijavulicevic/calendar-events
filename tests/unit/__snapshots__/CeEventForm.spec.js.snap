// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CeEventForm.vue should match snapshot 1`] = `
<div class="elevation-12 v-card v-sheet theme--light">
  <header class="v-sheet v-sheet--tile theme--dark v-toolbar v-toolbar--flat primary">
    <div class="v-toolbar__content" style="height: 64px;">
      <div class="v-toolbar__title">Create event</div>
      <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close theme--dark"></i></span></button>
    </div>
  </header>
  <div class="v-card__text">
    <form class="v-form" novalidate="novalidate">
      <div class="v-input theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-113" class="v-label theme--light" style="left: 0px; position: absolute;">Name</label><input data-vv-name="calendarEvent.name" data-vv-as="name" required="required" id="input-113" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
      <div class="layout justify-space-between">
        <div class="flex md5">
          <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field">
            <div class="v-input__prepend-outer">
              <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
            </div>
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-118" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Start date</label><input id="input-118" readonly="readonly" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <div class="v-menu">
            <div role="menu" class="v-menu__content theme--light " style="max-height: auto; min-width: 290px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
          </div>
        </div>
        <div class="flex md5">
          <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field">
            <div class="v-input__prepend-outer">
              <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
            </div>
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-126" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">End date</label><input id="input-126" readonly="readonly" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <div class="v-menu">
            <div role="menu" class="v-menu__content theme--light " style="max-height: auto; min-width: 290px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
          </div>
        </div>
      </div>
      <!---->
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox accent--text">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-input--selection-controls__input"><input aria-checked="true" id="input-132" role="checkbox" type="checkbox" value="">
              <div class="v-input--selection-controls__ripple accent--text"></div><i aria-hidden="true" class="v-icon notranslate material-icons theme--light accent--text">$vuetify.icons.checkboxOn</i>
            </div><label for="input-132" class="v-label theme--light" style="left: 0px; position: relative;">All day event</label>
          </div>
          <div class="v-messages theme--light accent--text"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
        </div>
      </div>
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-136" class="v-input__slot">
            <div class="v-select__slot"><label for="input-136" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Visibility</label>
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">Private</div><input id="input-136" readonly="readonly" type="text" aria-readonly="true">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$vuetify.icons.dropdown</i></div>
              </div>
            </div>
            <div class="v-menu">
              <div class="v-menu__content theme--light " style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
      <div class="v-input v-textarea theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-143" class="v-label theme--light" style="left: 0px; position: absolute;">Details</label><textarea data-vv-name="calendarEvent.details" data-vv-as="details" id="input-143" rows="3"></textarea></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
      <div class="v-input theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-146" class="v-input__slot">
            <div class="v-select__slot"><label for="input-146" class="v-label theme--light" style="left: 0px; position: absolute;">Participants</label>
              <div class="v-select__selections"><input id="input-146" readonly="readonly" type="text" aria-readonly="true"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$vuetify.icons.dropdown</i></div>
              </div>
            </div>
            <div class="v-menu">
              <div class="v-menu__content theme--light " style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
      <div class="v-input theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-153" class="v-label theme--light" style="left: 0px; position: absolute;">Location</label><input data-vv-name="calendarEvent.location" data-vv-as="location" required="required" id="input-153" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div class="v-card__actions">
    <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default blue--text text--darken-1"><span class="v-btn__content">Create</span></button>
  </div>
</div>
`;
